CodeEdit/Features/Editor/AutoCompleteCoordinator.swift

```diff
@@ -0,0 +1,156 @@
+//
+//  AutoCompleteCoordinator.swift
+//  CodeEdit
+//
+//  Created by Abe Malla on 9/20/24.
+//
+
+import AppKit
+import CodeEditTextView
+import CodeEditSourceEditor
+import LanguageServerProtocol
+
+class AutoCompleteCoordinator: TextViewCoordinator {
+    /// A reference to the `TextViewController`, to be able to make edits
+    private weak var textViewController: TextViewController?
+    /// A reference to the file we are working with, to be able to query file information
+    private unowned var file: CEWorkspaceFile
+    /// The event monitor that looks for the keyboard shortcut to bring up the autocomplete menu
+    private var localEventMonitor: Any?
+    /// The `ItemBoxWindowController` lets us display the autocomplete items
+    private var itemBoxController: ItemBoxWindowController?
+
+    init(_ file: CEWorkspaceFile) {
+        self.file = file
+    }
+
+    func prepareCoordinator(controller: TextViewController) {
+        itemBoxController = ItemBoxWindowController()
+        itemBoxController?.delegate = self
+        itemBoxController?.close()
+        self.textViewController = controller
+
+        localEventMonitor = NSEvent.addLocalMonitorForEvents(matching: .keyDown) { event in
+            // `ctrl + space` keyboard shortcut listener for the item box to show
+            if event.modifierFlags.contains(.control) && event.charactersIgnoringModifiers == " " {
+                Task {
+                    await self.showAutocompleteWindow()
+                }
+                return nil
+            }
+            return event
+        }
+    }
+
+    /// Will query the language server for autocomplete suggestions and then display the window.
+    @MainActor
+    func showAutocompleteWindow() {
```

Comment:
I mentioned it in the review of the CETV pr, but this function should be moved to CESE. That should simplify this type, as you could instead add a parameter to CESE for an optional `ItemBoxDelegate`, and ignore the unnecessary `TextViewCoordinator` conformance. Then make the `ItemBoxDelegate` pass the text controller when it inserts items so you can still make that range conversion on line 111.

That does make it a little harder to get this object down to be visible to the `CodeFileView`, but check how it works with the content coordinator. I tried to make it so it's extendable for new types like this.
- Add a published property to CodeFileDocument
- Extend the type in `LanguageServerFileMap` to contain this object
- When opening a document, set this object in the `CodeFileDocument`'s new published property